dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconController.java

```diff
@@ -66,70 +77,43 @@
 @OpenApi.Tags( "ui" )
 @Controller
 @RequestMapping( value = IconSchemaDescriptor.API_ENDPOINT )
+@RequiredArgsConstructor
 @ApiVersion( { DhisApiVersion.DEFAULT, DhisApiVersion.ALL } )
 public class IconController
 {
     private static final int TTL = 365;
 
-    @Autowired
-    private IconService iconService;
-
-    @Autowired
-    private ContextService contextService;
-
-    @GetMapping
-    public @ResponseBody List<IconData> getIcons( HttpServletResponse response,
-        @RequestParam( required = false ) Collection<String> keywords )
-    {
-        Collection<IconData> icons;
-
-        if ( keywords == null )
-        {
-            icons = iconService.getIcons();
-        }
-        else
-        {
-            icons = iconService.getIcons( keywords );
-        }
+    private final IconService iconService;
 
-        return icons.stream()
-            .map( data -> data.setReference( String.format( "%s%s/%s/icon.%s", contextService.getApiPath(),
-                IconSchemaDescriptor.API_ENDPOINT, data.getKey(), Icon.SUFFIX ) ) )
-            .collect( Collectors.toList() );
-    }
+    private final FileResourceService fileResourceService;
 
-    @GetMapping( "/keywords" )
-    public @ResponseBody Collection<String> getKeywords( HttpServletResponse response )
-    {
-        return iconService.getKeywords();
-    }
+    private final IconMapper iconMapper;
 
     @GetMapping( "/{iconKey}" )
-    public @ResponseBody IconData getIcon( HttpServletResponse response, @PathVariable String iconKey )
-        throws WebMessageException
+    public @ResponseBody IconResponse getIcon( @PathVariable String iconKey )
+        throws WebMessageException,
+        NotFoundException
     {
-        Optional<IconData> icon = iconService.getIcon( iconKey );
+        Icon icon = iconService.getIcon( iconKey );
 
-        if ( !icon.isPresent() )
+        if ( icon == null )
         {
             throw new WebMessageException(
                 notFound( String.format( "Icon not found: '%s", iconKey ) ) );
         }
 
-        icon.get().setReference( String.format( "%s%s/%s/icon.%s", contextService.getApiPath(),
-            IconSchemaDescriptor.API_ENDPOINT, icon.get().getKey(), Icon.SUFFIX ) );
-
-        return icon.get();
+        return iconMapper.from( icon );
     }
 
     @GetMapping( "/{iconKey}/icon.svg" )
     public void getIconData( HttpServletResponse response, @PathVariable String iconKey )
         throws WebMessageException,
-        IOException
+        IOException,
+        NotFoundException
     {
         Optional<Resource> icon = iconService.getIconResource( iconKey );
 
-        if ( !icon.isPresent() )
+        if ( icon.isEmpty() )
```

Comment:
this should also not be necessary once `iconService.getIconResource` throws NotFoundException (see comment in service)